### PR TITLE
[6.x] Override (static) password name

### PR DESCRIPTION
**Pull request by ahmedsayedabdelsalam**
There was a case when I needed to make the User model store the password in a different column rather than "password", unfortunately, this column can't be changed, so to make it more dynamic, I've added a method that can be overridden, where we can specify the name of the column that stores the password, and an attribute in the User model that has a default value for that column with a value = "password"

old: if (Str::contains($key, 'password'))

new: if (Str::contains($key, $query->getModel()->getPasswordName()))

the new attribute: protected $passwordName = 'password';

### CODE REVIEW

File: src/Illuminate/Auth/GenericUser.php [hunk 1]
```diff
@@ -131,4 +131,14 @@ public function __unset($key)
     {
         unset($this->attributes[$key]);
     }
+
+    /**
+     * Get the column name for the "password".
+     *
+     * @return string
+     */
+    public function getPasswordName()
```
Comment (jmarcher) [inline src/Illuminate/Auth/GenericUser.php]: Why are you duplicating this here?

**Review by jmarcher — CHANGES_REQUESTED** [review]

**Comment by jmarcher** [comment]
You are missing at least one case: https://github.com/laravel/framework/blob/82af288b24a3abf454ed55a3147dd873f56a133b/src/Illuminate/Auth/GenericUser.php#L56

**Comment by iamthefreeman** [comment]
Should this be targeted to master as this has breaking changes in here?

**Comment by MattStrauss** [comment]
I [commented on the issue](https://github.com/laravel/framework/issues/30160#issue-501507501) before realizing there was a PR for it. Perhaps my line of thinking is off-base with this, but if not, then a PR may not be necessary. 

https://github.com/laravel/framework/issues/30160#issue-501507501

File: src/Illuminate/Auth/EloquentUserProvider.php [hunk 2]
```diff
@@ -117,7 +117,7 @@ public function retrieveByCredentials(array $credentials)
         $query = $this->newModelQuery();
 
         foreach ($credentials as $key => $value) {
-            if (Str::contains($key, 'password')) {
+            if (Str::contains($key, $query->getModel()->getPasswordName())) {
```
Comment (X-Coder264) [inline src/Illuminate/Auth/EloquentUserProvider.php]: There's a `array_key_exists('password', $credentials)` check just a few lines above this one.

File: src/Illuminate/Contracts/Auth/Authenticatable.php [hunk 3]
```diff
@@ -46,4 +46,11 @@ public function setRememberToken($value);
      * @return string
      */
     public function getRememberTokenName();
+
+    /**
+     * Get the column name for the "password".
+     *
+     * @return string
+     */
+    public function getPasswordName();
```
Comment (X-Coder264) [inline src/Illuminate/Contracts/Auth/Authenticatable.php]: Adding a method to an interface is a breaking change so this PR should target `master`.

File: src/Illuminate/Auth/GenericUser.php (hunk 1 above)
Comment (X-Coder264) [reply src/Illuminate/Auth/GenericUser.php]: This is here because he added this method to the interface which this generic user implements.

**Comment by GrahamCampbell** [comment]
Yes, please send this to master (7.x).

**Comment by MattStrauss** [comment]
@ahmedsayedabdelsalam 

I just tried out the idea from [this comment](https://github.com/laravel/framework/issues/30160#issuecomment-537894304) on my local machine and it works as expected. So essentially you can just override the `getAuthPassword` method on your `User` model and just change the name of the property it is accessing to match the column in your DB without any other changes to the Framework. 

Here's a quick example:

`User Model`
```
    // here we assume the password column is set to "password2" as opposed to the normal "password"  
    public function getAuthPassword()
    {
        return $this->password2;
    }
```

**Comment by ahmedsayedabdelsalam** [comment]
> @ahmedsayedabdelsalam
> 
> I just tried out the idea from [this comment](https://github.com/laravel/framework/issues/30160#issuecomment-537894304) on my local machine and it works as expected. So essentially you can just override the `getAuthPassword` method on your `User` model and just change the name of the property it is accessing to match the column in your DB without any other changes to the Framework.
> 
> Here's a quick example:
> 
> `User Model`
> 
> ```
>     // here we assume the password column is set to "password2" as opposed to the normal "password"  
>     public function getAuthPassword()
>     {
>         return $this->password2;
>     }
> ```

Thanks for your reply and excuse me for the late response
that's a long time :D I can't remember actually what was the situation needed me to use the following function 🤦‍♂️
but the problem was that this function uses the password column statically so
   ```
 /**
     * Retrieve a user by the given credentials.
     *
     * @param  array  $credentials
     * @return \Illuminate\Contracts\Auth\Authenticatable|null
     */
    public function retrieveByCredentials(array $credentials)
    {
        if (empty($credentials) ||
           (count($credentials) === 1 &&
            array_key_exists('password', $credentials))) {
            return;
        }`

        // First we will add each credential element to the query as a where clause.
        // Then we can execute the query and, if we found a user, return it in a
        // Eloquent User "model" that will be utilized by the Guard instances.
        $query = $this->newModelQuery();

        foreach ($credentials as $key => $value) {
            if (Str::contains($key, $query->getModel()->getPasswordName())) {
                continue;
            }

            if (is_array($value) || $value instanceof Arrayable) {
                $query->whereIn($key, $value);
            } else {
                $query->where($key, $value);
            }
        }

        return $query->first();
    }
```

so I think 
```
public function getAuthPassword()
    {
        return $this->password2;
    }
``` 
with not help here.

**Pull request closed without merging**